at86rf2xx: release framebuffer on recv with (len > 0) && (buf == NULL)

Josar opened the pull request:
The upper layer calls the recieve function to get back the size of the frame. 
At the moment the frame buffer is released when returning the size, but this should not be done. 
see https://github.com/RIOT-OS/RIOT/pull/9172#discussion_r200327609

This PR sets the tranceiver in `PLL_ON` state to avoid corruption of the data in the frame buffer and sets it back to the the last state which the transceiver had before changing into transmit mode after the data is read out.

### Issues/PRs references
#9172

gebart commented on drivers/at86rf2xx/at86rf2xx_netdev.c:
```diff
@@ -145,7 +145,6 @@ static int _recv(netdev_t *netdev, void *buf, size_t len, void *info)
 
     /* just return length when buf == NULL */
     if (buf == NULL) {
-        at86rf2xx_fb_stop(dev);
```
This should be if (len > 0), instead of removed

Josar commented:
@gebart got that covered.

miri64 commented:
@Josar for future reference, can you please provide your changes as fixup commits, so your changes can be tracked during the review? Thank you!

Josar commented:
I tried to find the proper way of handling this and used Atmel-8351E-MCU_Wireless-AT86RF233_Datasheet_072014 as reference to get some ideas. All page number refer to this if not stated otherwise.
 
page 170 Dynamic Frame Buffer Protection.
The SPI release indicated by /SEL = H is what releases the frame buffer protection.

I can think of this possible solutions.

1. Using Dynamic FBP (page 170) but Read PHR after IRQ_2 (RX_START) (page 24) with SRAM acces and not framebuffer acces. I think this can also be done in `_rec` which is a while after IRQ_2. 
But as state in AT86RF231 manual page 23 "The frame length information can be accessed only
using Frame Buffer access.". So this seems not possible for all at86rx2xx.

1.  Using Dynamic FBP (page 170)  call `at86rf2xx_fb_start` in the `if` when returning the size and keep the spi bus locked. Then read out the data when the second call to `_rev` is executed. And unlock the spi after read out is finished.

This would ensure that Dynamic frame buffer protection works and the radio transceiver automatically enters the power save mode (page 174). I think in both cases 1 and 2.

1. set `Static Frame Buffer Protection`  with `RX_PDT_DIS` bit of `RX_SYN`. Manual page 112

1. Set receiver state from `RX_AACK_ON` to `PLL_ON` directly after IRQ_3 (TRX_END) (page 118)

Which one (2,3,4) do you think is the best ore would you prefer?

kYc0o commented:
I tested it and it actually breaks the at86rf2xx driver.

Josar commented:
Implemented solution 4 could someone with access to hardware make some test?

gebart commented:
Solution 4 sounds like the best option in theory. I am in holidays and will not be able to test this PR.

kYc0o commented:
> Implemented solution 4 could someone with access to hardware make some test?

I'm on it.

kYc0o commented:
It works again. @gebart do you ACK?

kYc0o commented:
I think @gebart comments were addressed, and since he's not available I'd say we can dismiss his review. @miri64 what do you think?

miri64 commented:
Maybe this question reveals why my opinion shouldn't be asked here ;-), but in `at86rf2xx_internal.c` there are still two occurences of `AT86RF2XX_STATE_TRX_OFF` that as far as I understand might conflict with the changed idle state here. As far as my testings go here though, it works for me.

kYc0o commented:
Ok, maybe @bergzand can look at it.

Josar commented:
@miri64 i think the name of `idle_state` is missleading. This variable is used to save the operation state when changing to transmit. So the old state is saved and after transmission returned to the old `idle_state`. It is now also used when receiving in the same manner return to the last state before transmission.
I just changed the initial state to match the state after reset, so if there is a reception after the reset it will return to the right state.
I changed it to be done it in the `at86rf2xx_reset`, but i thought this might be redundant.

Also `idle_state` is only changed when reseting and before transmission.

The other `AT86RF2XX_STATE_TRX_OFF` you refer to are states which have to be taken when changing from TX to RX and vice versa or when configuring some options. imho.

Josar commented:
@gebart maybe you could have a second look.

Josar commented:
@kYc0o @gebart would love to get this done and continue with https://github.com/RIOT-OS/RIOT/pull/9172

kYc0o commented:
Thanks for your review and test @gebart. @Josar you may squash.

Josar commented:
@kYc0o rebased and squashed.

kYc0o commented:
Some fixup commit appeared. Murdock won't pass until all fixup commits are squashed.

Josar commented:
I changed sequence as the state of the transceiver can not be change after the SPI is release, imho.
please have a second look. I can not test this an hardware.

Josar commented:
@kYc0o  i know my last change was wrong so i reverted it. Could you test it again.

kYc0o commented:
> But i don`t think this should have an effect to the observed problem.

True, this PR somehow adds a delay between releasing the device and be ready to receive the next packet.

I'm afraid then this behaviour needs more study to catch the real problem.

kYc0o commented on drivers/at86rf2xx/at86rf2xx_netdev.c:
```diff
@@ -143,11 +146,21 @@ static int _recv(netdev_t *netdev, void *buf, size_t len, void *info)
+        if (len > 0) {
+            /* set device back in operation state which was used before last transmission.
+             * e.g RX_AACK_ON */
+            at86rf2xx_set_state(dev, dev->idle_state);
```
I'm not really sure about this state, who sets it (e.g. `RX_AACK_ON`)?

Josar replied:
https://github.com/RIOT-OS/RIOT/blob/4c753ece542951d0ba4313a3f914972ffe4f228b/drivers/at86rf2xx/at86rf2xx.c#L112

It is set her and when starting to transmit, then the old state is saved.

Josar commented:
I have `ping timeouts` on master with the atmega256rfr2 and a atxmega256 with at86rf233 when sending bigger payload. 
With the jiminy i have a 98% rate. And with the other board it is 75%.
One difference i found is that the `Acknowledge Request` Flag in the 802.15.4 datagram is not set for the xmega board. I don't know why it is disabled and how to enable in, any one a suggestion?

Josar commented:
So activating `NETDEV_IEEE802154_ACK_REQ` got me 100% received.

I am not sure when this changed and where and if it was on purpose or not.
But in privious builds it was active and now it is not anymore.

Josar commented:
@kYc0o can you spot a difference when observing the packets with wireshark?

Josar commented:
@kYc0o please check if the ack_req flag is set see https://github.com/RIOT-OS/RIOT/issues/9957 .
If not set it with `ifconfig 7 ack_req` and test again. For me this resulted in pretty good pings.

Josar commented:
@kYc0o did you have time to check the `ack_req`?

kYc0o commented:
@Josar I'll rebase my local branch on master to test, would you mind to rebase yours?

kYc0o commented:
Everything OK with #10250 in. Please squash and rebase.

jia200x commented:
This needs to be backported to the `2018.04-branch` as well. I will add the label.

jia200x commented:
@Josar thank you for your contribution! Could you open a PR against `2018.10-branch` to merge this bugfix in the Release?

Josar commented:
@jia200x see https://github.com/RIOT-OS/RIOT/pull/10285

But while rebasing for https://github.com/RIOT-OS/RIOT/pull/9172/ i had some second thoughts about setting the right mode after reading out the buffer.

As it is now it is hard to understand where the idle state is saved.
